db/sqldb/owner.go

```diff
@@ -0,0 +1,167 @@
+package sqldb
+
+import (
+	"fmt"
+
+	"github.com/RMI/pacta/db"
+	"github.com/RMI/pacta/pacta"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgtype"
+)
+
+const ownerIDNamespace = "own"
+const ownerSelectColumns = `
+	owner.id,
+	owner.user_id,
+	owner.initiative_id
+`
+
+func (d *DB) Owners(tx db.Tx, ids []pacta.OwnerID) (map[pacta.OwnerID]*pacta.Owner, error) {
+	ids = dedupeIDs(ids)
+	rows, err := d.query(tx, `
+		SELECT `+ownerSelectColumns+`
+		FROM owner
+		WHERE owner.id IN `+createWhereInFmt(len(ids))+`;`, idsToInterface(ids)...)
+	if err != nil {
+		return nil, fmt.Errorf("querying owners: %w", err)
+	}
+	return rowsToOwners(rows)
+}
+
+func (d *DB) Owner(tx db.Tx, id pacta.OwnerID) (*pacta.Owner, error) {
+	owners, err := d.Owners(tx, []pacta.OwnerID{id})
+	if err != nil {
+		return nil, fmt.Errorf("querying owner: %w", err)
+	}
+	return exactlyOneFromMap("owner", id, owners)
+}
+
+func (d *DB) ownerByUser(tx db.Tx, id pacta.UserID) (*pacta.Owner, error) {
+	rows, err := d.query(tx, `SELECT `+ownerSelectColumns+` FROM owner WHERE owner.user_id = $1;`, id)
+	if err != nil {
+		return nil, fmt.Errorf("querying owners: %w", err)
+	}
+	owners, err := rowsToOwners(rows)
+	if err != nil {
+		return nil, fmt.Errorf("translating rows to owners: %w", err)
+	}
+	for _, owner := range owners {
+		if owner.User != nil && owner.User.ID == id {
+			return owner, nil
+		}
+	}
+	return nil, db.NotFound(id, "ownerByUserId")
+}
+
+func (d *DB) ownerByInitiative(tx db.Tx, id pacta.InitiativeID) (*pacta.Owner, error) {
+	rows, err := d.query(tx, `SELECT `+ownerSelectColumns+` FROM owner WHERE owner.initiative_id = $1;`, id)
+	if err != nil {
+		return nil, fmt.Errorf("querying owners: %w", err)
+	}
+	owners, err := rowsToOwners(rows)
+	if err != nil {
+		return nil, fmt.Errorf("translating rows to owners: %w", err)
+	}
+	for _, owner := range owners {
+		if owner.Initiative != nil && owner.Initiative.ID == id {
+			return owner, nil
+		}
+	}
+	return nil, db.NotFound(id, "ownerByInitiativeId")
+}
+
+func (d *DB) GetOrCreateOwnerForUser(tx db.Tx, uID pacta.UserID) (pacta.OwnerID, error) {
+	owner, err := d.ownerByUser(tx, uID)
+	if err == nil {
+		return owner.ID, nil
+	}
+	if !db.IsNotFound(err) {
+		return "", fmt.Errorf("querying owner by user: %w", err)
+	}
+	return d.createOwner(tx, &pacta.Owner{User: &pacta.User{ID: uID}})
+}
+
+func (d *DB) GetOrCreateOwnerForInitiative(tx db.Tx, iID pacta.InitiativeID) (pacta.OwnerID, error) {
+	owner, err := d.ownerByInitiative(tx, iID)
+	if err == nil {
+		return owner.ID, nil
+	}
+	if !db.IsNotFound(err) {
+		return "", fmt.Errorf("querying owner by user: %w", err)
+	}
+	return d.createOwner(tx, &pacta.Owner{Initiative: &pacta.Initiative{ID: iID}})
```

Comment:
This feels like a DB endpoint that would benefit from a `StartOrContinueTransaction` block or whatever we call it now

Reply:
Good call.